File: src/lib/tauri/events.ts
import { listen, type UnlistenFn } from '@tauri-apps/api/event'
import type { TaskEventPayload } from '@/lib/tauri/contracts'

export const APP_READY_EVENT = 'app:ready'
export const TASK_PROGRESS_EVENT = 'task:progress'
export const TASK_COMPLETED_EVENT = 'task:completed'
export const TASK_FAILED_EVENT = 'task:failed'

export async function listenTaskEvents(handler: (payload: TaskEventPayload) => void) {
	const unlistenProgress = await listen<TaskEventPayload>(TASK_PROGRESS_EVENT, (event) => handler(event.payload))
	const unlistenCompleted = await listen<TaskEventPayload>(TASK_COMPLETED_EVENT, (event) => handler(event.payload))
	const unlistenFailed = await listen<TaskEventPayload>(TASK_FAILED_EVENT, (event) => handler(event.payload))

	return () => {
		unlistenProgress()
		unlistenCompleted()
		unlistenFailed()
	}
}

export async function listenAppReady(handler: (payload: unknown) => void): Promise<UnlistenFn> {
	return listen(APP_READY_EVENT, (event) => handler(event.payload))
}

File: src-tauri/src/app/events.rs
use tauri::{AppHandle, Emitter};

use crate::models::dto::TaskEventPayload;

pub const APP_READY_EVENT: &str = "app:ready";
pub const TASK_PROGRESS_EVENT: &str = "task:progress";
pub const TASK_COMPLETED_EVENT: &str = "task:completed";
pub const TASK_FAILED_EVENT: &str = "task:failed";

pub fn emit_app_ready(app: &AppHandle, payload: serde_json::Value) {
    let _ = app.emit(APP_READY_EVENT, payload);
}

pub fn emit_task_progress(app: &AppHandle, payload: &TaskEventPayload) {
    let _ = app.emit(TASK_PROGRESS_EVENT, payload);
}

pub fn emit_task_completed(app: &AppHandle, payload: &TaskEventPayload) {
    let _ = app.emit(TASK_COMPLETED_EVENT, payload);
}

#[allow(dead_code)]
pub fn emit_task_failed(app: &AppHandle, payload: &TaskEventPayload) {
    let _ = app.emit(TASK_FAILED_EVENT, payload);
}
